lib/stat_tracker.rb

```diff
@@ -14,34 +21,309 @@ def self.from_csv(data_files)
+    total_games = combine_all_games_played(team_id).each do |game|
+      if team_id == game.home_team_id && game.outcome == :home_win
+        wins_and_ties += 1.0
+      elsif team_id == game.away_team_id && game.outcome == :away_win
+        wins_and_ties += 1.0
+      end
+    end
+    (wins_and_ties / total_games.count).round(2)
+  end
+
+  def average_win_percentage(team_id)
+    find_total_wins_or_ties(team_id)
+  end
+
+  def most_home_goals_scored(team_id)
+    home_games_filtered_by_team(team_id).max_by do |game|
+      game.home_goals
+    end.home_goals.to_i
+  end
+
+  def most_away_goals_scored(team_id)
+    away_games_filtered_by_team(team_id).max_by do |game|
+      game.away_goals
+    end.away_goals.to_i
+  end
+
+  def most_goals_scored(team_id)
+    if most_home_goals_scored(team_id) > most_away_goals_scored(team_id)
+      most_home_goals_scored(team_id)
+    else
+      most_away_goals_scored(team_id)
+    end
+  end
+
+  def fewest_home_goals_scored(team_id)
+    home_games_filtered_by_team(team_id).min_by do |game|
+      game.home_goals
+    end.away_goals.to_i
+  end
+
+  def fewest_away_goals_scored(team_id)
+    away_games_filtered_by_team(team_id).min_by do |game|
+      game.away_goals
+    end.away_goals.to_i
+  end
+
+  def fewest_goals_scored(team_id)
+    if fewest_home_goals_scored(team_id) < fewest_away_goals_scored(team_id)
+      fewest_home_goals_scored(team_id)
+    else
+      fewest_away_goals_scored(team_id)
+    end
+  end
+
+  def all_games_played_by_team(team_id)
+    game_collection.all.find_all do |game|
+      game.home_team_id == team_id || game.away_team_id == team_id
+    end
+  end
+
+  def opponents(team_id)
+    # We want:
+                # {
+                #   "some_opponent_id" =>
+                #     {
+                #       "won" => times_some_opponent_won_against_my_team,
+                #       "lost" => times_some_opponent_lost_against_my_team,
+                #       "tied" => times_some_opoonent_tied_against_my_team
+                #     }
+                # }
+    acc = Hash.new
+    all_games_played_by_team(team_id).each do |game|
+      if game.away_team_id == team_id
+        opponent_id = game.home_team_id
+      else
+        opponent_id = game.away_team_id
+      end
+
+      # if the block has been iterated through before, then set old_value won/lost/tied to that previous value, and if it has not been iterated through then set old value won/lost/tied to 0
+
+      if acc[opponent_id].nil?
+        old_value_won = 0.0
+        old_value_lost = 0.0
+        old_value_tied = 0.0
+      else
+        old_value_won = acc[opponent_id][:won]
+        old_value_lost = acc[opponent_id][:lost]
+        old_value_tied = acc[opponent_id][:tied]
+      end
+
+      if opponent_id == game.winning_team_id
+        old_value_won += 1.0
+      elsif opponent_id == game.losing_team_id
+        old_value_lost += 1.0
+      else
+        old_value_tied += 1.0
       end
+
```

Comment:
Very clever use of hashes here to achieve the desired output of this method. Seems like it was very complicated to figure out! Kudos!